### ISSUE TITLE
[Proposal] Prevent express middleware send response

### ISSUE THREAD

**Issue by chemitaxis** [zenstackhq/zenstack#531]
Hello, the middleware functions should primarily be responsible for processing the request and preparing the response, but not actually sending it. Instead, once they have done their part in processing, they should pass the control to the next middleware in line using the next() function. 

This ensures a proper flow of control in the application, allowing other middleware functions or the main request handler to perform additional actions if necessary.

For instance, you may have middleware functions for tasks like data validation, logging, handle error, etc. After these tasks are completed, the response is then generated and sent in the main request handler, providing a central location for response generation.

This approach separates concerns more effectively, making the codebase more modular and maintainable. It also allows more flexibility for further processing of the response before it is sent. Plus, error handling can be more centralized and robust, as you can have specific error-handling middleware.

For testing, this design makes it much easier, as each middleware function can be tested independently for its specific task. Testing becomes straightforward as you just need to ensure that each middleware behaves correctly when receiving a request and calling next().

This way, we keep up with the principle of each part of the software doing one thing and doing it well. The middleware processes the request and prepares the response, the main request handler (or other middleware) sends the response, and the code becomes more maintainable.

What do you think?

**Comment by ymc9** [comment]
Hi @chemitaxis , thanks for the detailed explanation I think it all makes very good sense.

For simple cases, I think it's easier to use if the middleware just directly sends the response. Shall we add an option to the middleware factory to control this behavior?

```ts
app.use(
    '/api/model',
    ZenStackMiddleware({
        getPrisma: (request) => withPresets(prisma, { user: getSessionUser(request) }),
        sendResponse: false
    })
);
```

When `sendResponse` is false, the prepared response object can be stored into `res.locals` (and then call `next()`) and be used by the subsequent middleware or request handlers:

```ts
app.all('/api/model/*', (req, res) => {
    const { status, body } = res.locals.zenstack;
    // further processing
    ...
    res.status(status).json(body);
});
```

Does this sound reasonable to you?

**Comment by chemitaxis** [comment]
This sounds perfect. I can do another PR if you want

**Comment by ymc9** [comment]
> This sounds perfect. I can do another PR if you want

That'll be great! Thank you!

**Comment by chemitaxis** [comment]
@ymc9 here you have :) https://github.com/zenstackhq/zenstack/pull/533

**Comment by ymc9** [comment]
Fixed by #533
